extensions/amp-story/1.0/test/test-amp-story-quiz.js

```diff
@@ -0,0 +1,134 @@
+ * @param {number} numPrompts
+ * @param {number} numOptions
+ */
+const populateQuiz = (win, quizElement, numPrompts = 1, numOptions = 4) => {
+  for (let i = 0; i < numPrompts; i++) {
+    const prompt = win.document.createElement('h1');
+    prompt.textContent = 'prompt';
+    quizElement.appendChild(prompt);
+  }
+
+  const option = win.document.createElement('option');
+  option.textContent = 'option';
+  for (let i = 0; i < numOptions; i++) {
+    quizElement.appendChild(option.cloneNode());
+  }
+};
+
+/**
+ * Populates the quiz with a prompt and three options
+ *
+ * @param {Window} win
+ * @param {AmpStoryQuiz} quiz
+ */
+const populateStandardQuizContent = (win, quizElement) => {
+  populateQuiz(win, quizElement);
+};
+
+describes.realWin(
+  'amp-story-cta-layer',
+  {
+    amp: {
+      runtimeOn: true,
+      extensions: ['amp-story:1.0'],
+    },
+  },
+  env => {
+    let win;
+    let ampStoryQuiz;
+
+    beforeEach(() => {
+      win = env.win;
+      const ampStoryQuizEl = win.document.createElement('amp-story-quiz');
+
+      win.document.body.appendChild(ampStoryQuizEl);
+      ampStoryQuiz = new AmpStoryQuiz(ampStoryQuizEl);
+    });
+
+    it('should take the html and put it in the shadow DOM', async () => {
+      populateStandardQuizContent(win, ampStoryQuiz.element);
+      ampStoryQuiz.buildCallback();
+      await ampStoryQuiz.layoutCallback();
+      expect(ampStoryQuiz.element).to.have.property('shadowRoot');
+      expect(ampStoryQuiz.getShadowRoot().children.length).to.equal(2);
+      expect(ampStoryQuiz.element.children).to.be.empty;
+    });
+
+    it('should structure the content in the shadow DOM', async () => {
+      populateStandardQuizContent(win, ampStoryQuiz.element);
+      ampStoryQuiz.buildCallback();
+      await ampStoryQuiz.layoutCallback();
+
+      const shadowRootChildren = ampStoryQuiz.getShadowRoot().children;
+      // check shadow root structure
+      expect(shadowRootChildren[0]).to.be.a('HTMLStyleElement');
+      expect(shadowRootChildren[1]).to.be.a('HTMLDivElement');
+
+      // check quiz container structure
+      expect(shadowRootChildren[1].children).to.not.be.empty;
+      const shadowQuizContent = shadowRootChildren[1].children;
+      expect(shadowQuizContent[0]).to.be.a('HTMLDivElement');
+      expect(shadowQuizContent[0]).to.have.class(
+        'i-amp-story-quiz-prompt-container'
+      );
+      expect(shadowQuizContent[1]).to.have.class(
+        'i-amp-story-quiz-option-container'
+      );
+
+      // check prompt container structure
+      expect(shadowQuizContent[0].children.length).to.equal(1);
+      expect(shadowQuizContent[0].children[0]).to.have.class(
+        'i-amp-story-quiz-prompt'
+      );
+
+      // check option container structure
+      expect(shadowQuizContent[1].childNodes.length).to.equal(4);
+      shadowQuizContent[1].childNodes.forEach(option => {
+        expect(option).to.have.class('i-amp-story-quiz-option');
+      });
```

Comment:
This test is very thorough, which is both good and bad. The coverage is fantastic, but it becomes harder to maintain: any change on the template, even if unrelated, will require updating these tests.

One alternative could be, for example:

replacing
```
      expect(shadowQuizContent[0].children.length).to.equal(1);
      expect(shadowQuizContent[0].children[0]).to.have.class(
        'i-amp-story-quiz-prompt'
      );
```

with 

```
      expect(quiz.querySelectorAll('.i-amp-story-quiz-prompt')).to.have.length(1);
```

The main benefit being that you still check that you have the prompt, but your test is no longer tied to the DOM structure, which rendering will be tests by visual diff tests.

Reply:
Thanks for the suggestion, this method will be less brittle